Add vtables to classes for cxx binaries w/o RTTI

tushar3q34 opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [X] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository.
- [X] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style).
- [X] I've documented every `RZ_API` function and struct this PR changes.
- [X] I've added tests that prove my changes are effective (required for changes to `RZ_API`).
- [ ] I've updated the [Rizin book](https://github.com/rizinorg/book) with the relevant information (if needed).

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

Analysed the data xrefs in constructor to find the vtables corresponding to each class.

**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots/videos. This is your time to re-check that everything works and that you covered all the edge cases -->

...

**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->

...

tushar3q34 committed:
cpp no rtti vtable analysis

wargio commented on librz/arch/no_rtti.c:
```diff
@@ -0,0 +1,121 @@
+#include <rz_core.h>
+#include <rz_util.h>
+
+typedef struct rz_taint_state_t {
+	HtSP /*<char *,char *>*/ *method_class_map; ///< maps methods to class name
+} RzNoRTTIState;
```
this is ok, but wouldn't make more sense to just pass the HtSP pointer directly? the usage of a struct in C makes sense only if you need to pass many variables to a function, but here you just have a struct with one element.

Also you can just define HtSP as
```c
HtSS *method_class_map;
```
Since HtSS is already a map<string, string>

tushar3q34 replied:
I have it here because I plan to use this later for inheritance. But yes for this PR, I think I can remove it.

wargio commented on test/db/cmd/cmd_avD:
```diff
@@ -602,4 +602,64 @@ EXPECT=<<EOF
+|       |   0x00400a0b      add   rdx, 0x08
+|       |   0x00400a0f      mov   rdx, qword [rdx]
+|       |   0x00400a12      mov   rdi, rax
+|       |   0x00400a15      call  rdx                                  ; Virtual Call : sym.non_virtual_thunk_to_Dog::_Dog_0x400ca5 / sym.Dog::_Dog_0x400c7a
```
unrelated to the PR, but since we have this info, will the decompiler use these symbols/hints? @notxvilka @thestr4ng3r

notxvilka replied:
Not at the current stage. It would require better integration. But in theory it's possible, yes.

wargio commented on librz/arch/no_rtti.c:
```diff
@@ -0,0 +1,121 @@
+			vtable.id = NULL;
+			vtable.addr = vtable_info->saddr;
+			vtable.offset = 0;
+			vtable.size = 0;
```
```suggestion
			RzAnalysisVTable vtable = { 0 };
			vtable.addr = vtable_info->saddr;
```

wargio replied:
Please initialize the variables in this way, so if we add something in the struct we do not need to manually initialize something and avoids undefined behavior

wargio commented on librz/arch/no_rtti.c:
```diff
@@ -0,0 +1,121 @@
+				function = rz_analysis_get_fcn_in(core->analysis, addr, RZ_ANALYSIS_FCN_TYPE_NULL);
+			}
+			if (function) {
+				bool found = false;
+				ht_sp_find(method_class_map, function->name, &found);
+				if (!found) { // since we only care about constructors in this use case
+					ht_sp_insert(method_class_map, function->name, rz_str_dup(class_name));
+				}
+			}
```
invert this with a continue.

wargio commented on librz/arch/no_rtti.c:
```diff
@@ -0,0 +1,121 @@
+			vtable.size = 0;
+			RVTableMethodInfo *vmeth;
+			rz_vector_foreach (&vtable_info->methods, vmeth) {
+				RzAnalysisMethod meth;
```
same here, initialize it with `{ 0 }`

wargio replied:
also can you please rename this into either `m` or `method` ?

wargio commented on librz/arch/no_rtti.c:
```diff
@@ -0,0 +1,121 @@
+					meth.real_name = fcn ? rz_str_dup(fcn->name) : rz_str_newf("virtual_%" PFMT64d, meth.vtable_offset);
+					meth.method_type = RZ_ANALYSIS_CLASS_METHOD_VIRTUAL;
+				} else {
+					RzAnalysisMethod exist_meth;
```
same here

wargio replied:
same here, initialize it with `{ 0 }` and rename

wargio reviewed (changes requested):
overall looks good. add a doxygen comment at line 7 to explain what this file implements.

```c
/** \file  no_rtti.c
 *   this file implements ....
```

wargio reviewed (approved):
LGTM.
@notxvilka & @Rot127 please have a look

notxvilka commented on librz/arch/no_rtti.c:
```diff
@@ -0,0 +1,130 @@
+			if (!function) {
+				function = rz_analysis_get_fcn_in(core->analysis, addr, RZ_ANALYSIS_FCN_TYPE_NULL);
+			}
+			if (function) {
```
You probably can flip this condition and just use `!function { continue; }`

notxvilka commented on librz/arch/no_rtti.c:
```diff
@@ -0,0 +1,130 @@
+			RVTableMethodInfo *vmethod;
+			rz_vector_foreach (&vtable_info->methods, vmethod) {
+				RzAnalysisMethod method = { 0 };
+				if (!rz_analysis_class_method_exists_by_addr(context->analysis, class_name, vmethod->addr)) {
```
This nesting is quite complicated. I recommend extracting internal loop into a separate function

Rot127 replied:
```suggestion
				if (!rz_analysis_class_method_exists_by_addr(analysis, class_name, vmethod->addr)) {
```

notxvilka reviewed (changes requested):
Overall, looks good. Please address minor style suggestions.
Moreover, it would be nice not only check x86 but also at least ARM64

Rot127 commented on librz/arch/no_rtti.c:
```diff
@@ -0,0 +1,130 @@
+				if (found) {
+					continue;
+				}
+				// since we only care about constructors in this use case
```
What does this comment mean here?

Rot127 commented on librz/arch/no_rtti.c:
```diff
@@ -0,0 +1,130 @@
+				if (!rz_analysis_class_method_exists_by_addr(context->analysis, class_name, vmethod->addr)) {
+					method.addr = vmethod->addr;
+					method.vtable_offset = vmethod->vtable_offset;
+					RzAnalysisFunction *fcn = rz_analysis_get_function_at(context->analysis, vmethod->addr);
```
```suggestion
					RzAnalysisFunction *fcn = rz_analysis_get_function_at(analysis, vmethod->addr);
```

Rot127 commented on librz/arch/no_rtti.c:
```diff
@@ -0,0 +1,130 @@
+					method.method_type = RZ_ANALYSIS_CLASS_METHOD_VIRTUAL;
+				} else {
+					RzAnalysisMethod exist_meth = { 0 };
+					if (rz_analysis_class_method_get_by_addr(context->analysis, class_name, vmethod->addr, &exist_meth) == RZ_ANALYSIS_CLASS_ERR_SUCCESS) {
```
```suggestion
					if (rz_analysis_class_method_get_by_addr(analysis, class_name, vmethod->addr, &exist_meth) == RZ_ANALYSIS_CLASS_ERR_SUCCESS) {
```

Rot127 replied:
Also, add a `rz_warn_if_reached` for the else case.

Rot127 commented on librz/arch/no_rtti.c:
```diff
@@ -0,0 +1,130 @@
+						rz_analysis_class_method_fini(&exist_meth);
+					}
+				}
+				rz_analysis_class_method_set(context->analysis, class_name, &method);
```
```suggestion
				rz_analysis_class_method_set(analysis, class_name, &method);
```

Rot127 commented on librz/arch/no_rtti.c:
```diff
@@ -0,0 +1,130 @@
+				rz_analysis_class_method_set(context->analysis, class_name, &method);
+				rz_analysis_class_method_fini(&method);
+			}
+			rz_analysis_class_vtable_set(core->analysis, class_name, &vtable);
```
```suggestion
			rz_analysis_class_vtable_set(analysis, class_name, &vtable);
```

Rot127 commented on librz/include/rz_analysis.h:
```diff
@@ -2284,6 +2284,7 @@ RZ_API void rz_analysis_rtti_msvc_recover_all(RVTableContext *vt_context, RzList
 RZ_API char *rz_analysis_rtti_itanium_demangle_class_name(RVTableContext *context, const char *name);
 RZ_API bool rz_analysis_rtti_itanium_print_at_vtable(RVTableContext *context, ut64 addr, RzOutputMode mode);
 RZ_API void rz_analysis_rtti_itanium_recover_all(RVTableContext *vt_context, RzList /*<RVTableInfo *>*/ *vtables);
+RZ_API void rz_analysis_no_rtti_analysis(RVTableContext *context, RzList /*<RVTableInfo *>*/ *vtables);
```
Missing type annotations

tushar3q34 replied:
Can you please explain what exactly is missing here?

notxvilka replied:
```
Type comment at <rizin/librz/arch/no_rtti.c:101:108> lacks angle brackets
```
You can always check the CI output yourself

tushar3q34 replied:
It wasn't at this line so I was a bit confused. It was actually missing at line 101

Rot127 replied:
> Can you please explain what exactly is missing here?

I mean `RZ_BORROW`, `RZ_NONNULL` etc.

tushar3q34 committed:
added arm64 tests

tushar3q34 commented:
I added arm64 tests too.
In the recent commit, I made following changes :
- Recursive data xrefs check till first constructor match
-  Moved class analysis in all analysis (`aaa`) command below because their were some dependencies that were getting analysed later

notxvilka reviewed (approved):
LGTM with one minor thing

Rot127 commented on librz/arch/no_rtti.c:
```diff
@@ -0,0 +1,170 @@
+}
+
+/**
+ * \param context Vtble context object
+ * \param vtables List of virtual tables
+ * \brief Runs non-rtti required analysis on binaries
+ */
+RZ_API void rz_analysis_no_rtti_analysis(RZ_NONNULL RVTableContext *context, RZ_NONNULL RzList /*<RVTableInfo *>*/ *vtables) {
```
Sorry, last request. Please write more expressive doxygen here.

Also, first the `\brief` then the `\param`
